src/module.ts

```diff
@@ -3067,72 +3079,97 @@ export function allocPtrArray(ptrs: usize[] | null): usize {
   assert(ASC_TARGET != Target.WASM64);
   var len = ptrs.length;
   var ptr = binaryen._malloc(len << 2);
-  var idx = ptr;
-  for (let i = 0, k = len; i < k; ++i) {
-    let val = ptrs[i];
-    binaryen.__i32_store(idx, <i32>val);
-    idx += 4;
+  if ((ASC_TARGET | 0) == Target.JS) {
+    binaryen.HEAPU32.set(ptrs, ptr >>> 2);
+  } else {
+    var idx = ptr;
+    for (let i = 0, k = len; i < k; ++i) {
+      let val = ptrs[i];
+      binaryen.__i32_store(idx, <i32>val);
+      idx += 4;
+    }
   }
   return ptr;
 }
 
 function stringLengthUTF8(str: string): usize {
   var len = 0;
   for (let i = 0, k = str.length; i < k; ++i) {
-    let u = str.charCodeAt(i);
-    if (u >= 0xD800 && u <= 0xDFFF && i + 1 < k) {
-      u = 0x10000 + ((u & 0x3FF) << 10) | (str.charCodeAt(++i) & 0x3FF);
-    }
+    let u = str.charCodeAt(i) >>> 0;
```

Comment:
Always in range [0, 0xFFFF]
```suggestion
    let u = str.charCodeAt(i);
```

Reply:
Yes, but forcing it to unsigned make it little bit faster for v8

Reply:
Oh? So the useful hint is probably is that it is an int, not a float?

Reply:
In general all smi(s) in v8 is signed i31 by default. But sometimes better have unsigned int due to compiler has much more powerful peephole opts for unsigned ints rather than signed.